### PR TITLE
fix for issue 24660

### PR DESCRIPTION
**Pull request by SteveAlexander**
This is a fix for https://github.com/flutter/flutter/issues/24660

The fix is kind of hacky, and targeted at allowing widgets with GlobalKeys to be used in a TabBarView.

There's also a test that demonstrates that when AutomaticKeepAliveClientMixin widgets are used in a TabBarView, their state gets mixed around. This means that https://github.com/flutter/flutter/pull/21350 did not fully fix this problem.

Thanks to @rrousselGit for the a demo of how TabBarView fails with KeepAlives.

So far, I've added two new tests, and the oddly-named `AlwaysKeepAliveWidget2`, rather than update the tests that are there. I've done this to show clearly what is fixed by this PR, and also what challenge remains unaddressed by this PR.

Fixed by this PR: The test "Skipping tabs with a GlobalKey child works" fails without the changes to `tabs.dart`.

Demonstrated not working correctly by this PR: The test "Skipping tabs with a KeepAlive child does not work" demonstrates the current incorrect behaviour, while showing what the correct behaviour would be in a comment.

As @rrousselGit pointed out in https://github.com/flutter/flutter/issues/24660, it's likely an overall solution to this is to have TabBarView refrain from changing the order of its children to simulate a "warp" animation effect. There are alternative ways to achieve this effect more at the presentation level, leaving the structure of widget children alone.

The warp animation effect is: with tabs `A B C D`, if I'm on `A` and select `C`, then I see an animation directly from `A` to `C` as if `B` isn't there. We might also consider whether such an effect is something we want to have.

In other words, the current implementation of "warping" causes some Flutter apps to behave incorrectly in unexpected ways. One option is to remove the "warping" effect — this would solve these problems — and reintroduce "warping" in a future release, in a way that doesn't suffer these side-effects. Alternatively... to make "warping" an option for now, so that those apps that need KeepAlives can do without.

### CODE REVIEW

**Comment by SteveAlexander** [comment]
Hi @dnfield! You worked on https://github.com/flutter/flutter/pull/21350. I'd welcome your feedback on this PR, and ideas for how we might move a fix forward.

**Comment by SteveAlexander** [comment]
tool_tests-linux failure is to do with `DevFSContent file`. Doesn't look related to this PR.

**Comment by SteveAlexander** [comment]
The error in `tool_tests-linux` is `test/devfs_test.dart: DevFSContent file`. I can't see how this has anything to do with what's changed in this PR.

How can I re-run these checks?

**Comment by rrousselGit** [comment]
> How can I re-run these checks?

Make a new commit

The following will do:

```sh
git commit --allow-empty -m "Trigger pipeline"
```

**Comment by burkard** [comment]
Your PR #24821 worked for me. I have a TabBarView w/ 3 widgets, each one with a GlobalKey; it was crashing due to the #24660 issue. Now, it jumps from A to C right away.

It's WORKING, thanks to your workaround! 

**If someone is reading this, trying to find a solution**: don't forget to insert "super.build(context);" in the build method, or it will break the keepAlive. 

Thanks, @SteveAlexander, hope they add this PR soon to the updates.

**Comment by haolsh** [comment]
yes!it worked.
it fix two bugs👍 
1: the tabview with 3 or more tabs when switch 1 to 3 the keepalive works not well
2: add global key to tab bugs

**Comment by my9074** [comment]
I hope this PR can be merged as soon as possible 👍

**Comment by adar2378** [comment]
> yes!it worked.
> it fix two bugs👍
> 1: the tabview with 3 or more tabs when switch 1 to 3 the keepalive works not well
> 2: add global key to tab bugs

Hi, Could you please show me a little bit of code of how to do that? I've used 3 different global keys for my 3 tab views and also added automatic keepalive mixinclient inevery tabbarview. But still if i switch from tab 1 from tab 3 it rebuilds & disposes tab 2 and and also disposes tab 1 :(

**Comment by pakyo-pan** [comment]
I hope this PR can be merged as soon as possible 👍

**Comment by dnfield** [comment]
I think @chunhtai has worked on some related issues to this and it may be fixed now.  @chunhtai can you take a look at this?

**Comment by diegotori** [comment]
Looks like @chunhtai did fix it in pull request #31978. However, this change **did not** make stable recently. Dunno when it'll appear there, but it wasn't in 1.5.

**Comment by chunhtai** [comment]
> @dnfield, looks like @chunhtai did fix it in pull request #31978. However, this change **did not** make stable recently. Dunno when it'll appear there, but it wasn't in 1.5.

It have been reverted and relanded several days ago due to build failure, it should be in the next dev release.

**Comment by dnfield** [comment]
I'm going to go ahead and close this out.  @SteveAlexander - if you think something is still missing from the fix that landed for this (https://github.com/flutter/flutter/pull/31978), please feel free to reopen this or open a new PR.

**Comment by dnfield** [comment]
The next dev release should be out any day now :)

**Comment by diegotori** [comment]
@dnfield how about the next stable. As someone trying to build a production app using flutter, it may not be wise to point the channel to dev. In my humble opinion, this should be escalated to stable once it soaks for an adequate amount of time in dev. It would be awful to wait months for a critical update like this one to make stable.

**Comment by dnfield** [comment]
Stable releases happen about once per quarter.  It's unfortunate that this fix missed the last stable - @Hixie would be able to say more about whether we can hotfix it back into stable.

For this particular fix, it would probably be fairly simple for you to locally patch it into your own copy of stable if that helps.  @Hixie may have better ideas on that.

**Comment by Hixie** [comment]
We literally just pushed a stable build yesterday so unfortunately this will be the longest possible time between landing and reaching stable. We do betas monthly, though, and frankly there's nothing about our betas that's really any less stable than our stable releases, so you should feel confident using that channel. That'll be available in a few weeks.

**Comment by diegotori** [comment]
Understood @Hixie. Hopefully the beta channel is a reliable substitute for stable if necessary.

**Comment by DreamBuddy** [comment]
Thanks very much ! This rp saved my life!

**Comment by haolsh** [comment]
> > yes!it worked.
> > it fix two bugs+1
> > 1: the tabview with 3 or more tabs when switch 1 to 3 the keepalive works not well
> > 2: add global key to tab bugs
> 
> Hi, Could you please show me a little bit of code of how to do that? I've used 3 different global keys for my 3 tab views and also added automatic keepalive mixinclient inevery tabbarview. But still if i switch from tab 1 from tab 3 it rebuilds & disposes tab 2 and and also disposes tab 1 :(
> 
> Edit: I also made changes following this in the file [1c0e111](https://github.com/flutter/flutter/commit/1c0e11183be4301a0cb96236561513d9caba07da)

just replace your flutter libary's tab.dart with this one

**Comment by DreamBuddy** [comment]
1.6.0 seems to be fixed
